src/zalsa.rs

```diff
@@ -186,21 +188,22 @@ impl Zalsa {
         {
             let jar_type_id = jar.type_id();
             let mut jar_map = self.jar_map.lock();
-            *jar_map
-            .entry(jar_type_id)
-            .or_insert_with(|| {
-                let index = IngredientIndex::from(self.ingredients_vec.len());
-                let ingredients = jar.create_ingredients(self, index);
+            let mut should_create = false;
+            let index = *jar_map.entry(jar_type_id).or_insert_with(|| {
+                should_create = true;
+                IngredientIndex::from(self.ingredients_vec.len())
+            });
+            if should_create {
+                let aux = JarAuxImpl(self, &jar_map);
+                let ingredients = jar.create_ingredients(&aux, index);
```

Comment:
There's a comment on `jar_map` that suggests that it's important to only access `ingredients_vec` while the lock on `jar_map` is held

https://github.com/MichaReiser/salsa/blob/38ad6555ebe4374bbc69022e005f4136a1839ad9/src/zalsa.rs#L127-L133

So moving it out to avoid the deadlock might not be entirely safe

Reply:
For example. What could happen now is that two calls end up with the same index because the new ingredient isn't pushed on the `ingredients_vec` until line 206, so that `self.ingredients_vec.len()` returns the same length.

Reply:
Okay, I'll fix this